drones/image_processing/normalization.py

```diff
@@ -0,0 +1,250 @@
+
+    Returns:
+    ----------
+    image : np.ndarray
+        gray image
+    """
+    return cv.cvtColor(image, cv.COLOR_BGR2GRAY)
+
+
+def histogram_equalization_gray(image: np.ndarray) -> np.ndarray:
+    """Apply histogram normalization on single channel. If image have more channels, it will return input
+
+    Parameters:
+    ----------
+    image : np.ndarray
+        image to be normalized(grayscale)
+
+    Returns:
+    ----------
+    image : np.ndarray
+        normalized version of input image
+    """
+    if len(image.shape) > 2:
+        return image
+    return cv.equalizeHist(image)
+
+
+def histogram_equalization_luminance(image: np.ndarray) -> np.ndarray:
+    """Apply histogram normalization on luminance channel, image should have 3 channels
+
+    Parameters:
+    ----------
+    image : np.ndarray
+        image to be normalized(3 channels)
+
+    Returns:
+    ----------
+    image : np.ndarray
+        normalized version of input image
+    """
+    # convert from RGB color-space to YCrCb
+    ycrcb_img = cv.cvtColor(image, cv.COLOR_BGR2YCrCb)
+
+    # equalize the histogram of the Y channel
+    ycrcb_img[:, :, 0] = cv.equalizeHist(ycrcb_img[:, :, 0])
+
+    # convert back to RGB color-space from YCrCb
+    return cv.cvtColor(ycrcb_img, cv.COLOR_YCrCb2BGR)
+
+
+def histogram_equalization_color(image: np.ndarray) -> np.ndarray:
+    """Apply histogram normalization on every image channel
+
+    Parameters:
+    ----------
+    image : np.ndarray
+        image to be normalized
+
+    Returns:
+    ----------
+    image : np.ndarray
+        normalized version of input image
+    """
+    # segregate color streams
+    b, g, r = cv.split(image)
+    h_b, bin_b = np.histogram(b.flatten(), 256, [0, 256])
+    h_g, bin_g = np.histogram(g.flatten(), 256, [0, 256])
+    h_r, bin_r = np.histogram(r.flatten(), 256, [0, 256])
+    # calculate cdf
+    cdf_b = np.cumsum(h_b)
+    cdf_g = np.cumsum(h_g)
+    cdf_r = np.cumsum(h_r)
+
+    # mask all pixels with value=0 and replace it with mean of the pixel values
+    cdf_m_b = np.ma.masked_equal(cdf_b, 0)
+    cdf_m_b = (cdf_m_b - cdf_m_b.min()) * 255 / (cdf_m_b.max() - cdf_m_b.min())
+    cdf_final_b = np.ma.filled(cdf_m_b, 0).astype("uint8")
+
+    cdf_m_g = np.ma.masked_equal(cdf_g, 0)
+    cdf_m_g = (cdf_m_g - cdf_m_g.min()) * 255 / (cdf_m_g.max() - cdf_m_g.min())
+    cdf_final_g = np.ma.filled(cdf_m_g, 0).astype("uint8")
+
+    cdf_m_r = np.ma.masked_equal(cdf_r, 0)
+    cdf_m_r = (cdf_m_r - cdf_m_r.min()) * 255 / (cdf_m_r.max() - cdf_m_r.min())
+    cdf_final_r = np.ma.filled(cdf_m_r, 0).astype("uint8")
+
+    # merge the images in the three channels
+    img_b = cdf_final_b[b]
+    img_g = cdf_final_g[g]
+    img_r = cdf_final_r[r]
+
+    img_out = cv.merge((img_b, img_g, img_r))
+    return img_out
+
+
+def image_centralization(image: np.ndarray) -> np.ndarray:
+    """Centralize image value. It should be used after apply zero_one norm
```

Comment:
Same suggestion about the exception here. You have to check if the zero-one norm was actually applied.

Reply:
in this case it is only suggestion, it will work without zero_norm, but output will be normalized

Reply:
In that case, I don't see a need for such suggestion in the docstring. If the logic of this function doesn't depend on image being normalized, there is no point. Instead, I would explain what will happen if you pass normalized image vs non-normalized image, just as you did in this thread. What's more, you have to include it in return value docstring too. Also, I would at least log info message if the centralization is being applied on normalized image or not.